README
====================
Python Fedora Module
====================

:Author: Toshio Kuratomi
:Date: 19 August 2015
:Version: 0.5.x

The Fedora module provides a python API for building `Fedora Services`_ and
clients that connect to them.  It has functions and classes that help to build
TurboGears_ applications and classes to make building clients of those
services much easier.

.. _`Fedora Services`: doc/service.html
.. _TurboGears: http://www.turbogears.org

.. contents::

-------
License
-------

This python module is distributed under the terms of the GNU Lesser General
Public License Version 2 or later.

------------
Dependencies
------------

``python-fedora`` requires the ``munch``, ``kitchen``, and ``requests`` python
modules.  It used to use ``pycurl``, but was updated to use ``requests`` as of
version ``0.3.32``.
The ``flask_fas_openid`` module requires the ``python-openid`` and 
``python-openid-teams`` modules.

----------
Installing
----------

``python-fedora`` is found in rpm form in Fedora proper.  Sometimes a new
version will be placed in the Fedora Infrastructure yum repository for testing
within Infrastructure before being released to the general public.  Installing
from the yum repository should be as easy as::

	# yum install python-fedora

If you want to install from a checkout of the development branch, follow these
procedures::

    # git clone https://github.com/fedora-infra/python-fedora.git
    # cd python-fedora
    # ./setup.py install

See the configuration notes in each section for information on configuring
your application after install.

---------------------------
Fedora Accounts Integration
---------------------------

We provide several modules that make connecting to the `Fedora Account
System`_ easier.

.. _`Fedora Account System`: https://fedorahosted.org/fas

General Purpose API
===================
The ``fedora.accounts.fas2`` module allows code to integrate with the `Fedora
Account System`_. It uses the JSON interface provided by the Account System
servre to retrieve information about users.

Note: This API is not feature complete. If you'd like to help add methods,
document the existing methods, or otherwise aid in development of this API
please contact us on the infrastructure list: infrastructure@lists.fedoraproject.org
or on IRC: lmacken, abadger1999, and ricky in #fedora-admin, irc.freenode.net.

Using the general API requires instantiating an ``AccountSystem`` object. You
then use methods on the ``AccountSystem`` to get and set information on the
people in the account system.

At the moment, there are only a few methods implemented. Full documentation on
these methods is available from the ``AccountSystem``'s docstrings from the
interpreter or, for instance, by running::

    $ pydoc fedora.accounts.fas2.AccountSystem

Here's an example of using the ``AccountSystem``::

	from fedora.accounts.fas2 import AccountSystem
	from fedora.client import AuthError,
x
	# Get an AccountSystem object.  All AccountSystem methods need to be
	# authenticated so you might as well give username and password here.
	fas = AccountSystem(username='foo', password='bar')

	people = fas.people_by_id()

TurboGears Interface
====================

The TurboGears_ interface also uses the JSON interface to the account system.
It provides a TurboGears_ ``visit`` and ``identity`` plugin so a TurboGears_
application can authorize via FAS. Since the plugin operates over JSON, it is
possible to use these plugins on hosts outside of Fedora Infrastructure as
well as within.  Remember, however, that entering your Fedora password on a
third party website requires you to trust that website. So doing things this
way is more useful for developers wanting to work on their apps outside of
Fedora Infrastructure than a general purpose solution for allowing Fedora
Users to access your web app. (SSL client certificates and OpenID are better
solutions to this problem but they are still being implemented in the FAS2
server.)

Configuring
-----------
To configure your TurboGears_ application, you need to set the following
variables in your pkgname/config/app.cfg file::

    fas.url='https://admin.fedoraproject.org/accounts/'
    visit.on=True
    visit.manager="jsonfas"
    identity.on="True"
    identity.failure_url="/login"
    identity.provider="jsonfas"

---------------
Fedora Services
---------------

``python-fedora`` provides several helper classes and functions for building a
TurboGears_ application that works well with other `Fedora Services`_.  the
`Fedora Services`_ documentation is the best place to learn more about these.

-----------------
TurboGears Client
-----------------
There is a module to make writing a client for our TurboGears services very
easy.  Please see the `client documentation`_ for more details

.. _`client documentation`: doc/client.rst

-----------------
Building the docs
-----------------

You'll need to install python-sphinx for this::
  yum install python-sphinx

Then run this command::
  python setup.py build_sphinx

------------
Translations
------------

The strings in python-fedora has mainly error messages.  These are translated
so we should make sure that translators are able to translate them when
necessary.  You will need babel, setuptools, and zanata-client to run these
commands::
  yum install babel setuptools zanata-client

Much information about using zanata for translations can be found in the
`zanata user's guide`_.  The information in this section is largely from
experimenting with the information in the `zanata client documentation`_

.. _`zanata user's guide`: http://zanata.readthedocs.org
.. ~`zanata client documentation`: http://zanata.readthedocs.org/en/latest/user-guide/client-configuration/

Updating the POT File
=====================

When you make changes that change the translatable strings in the package, you
should update the POT file.  Use the following distutils command (provided by
python-babel) to do that::
  ./setup.py extract_messages -o translations/python-fedora.pot
  zanata-cli push

Then commit your changes to source control.

Updating the PO Files
=====================

fedora.zanata.org will merge the strings inside the pot file with the already
translated strings.  To merge these, we just need to pull revised versions of
the po files::
  zanata-cli pull

Then commit the changes to source control (look for any brand new PO files that
zanata may have created).

Creating a new PO File
======================

The easiest way to create a new po file for a new language is in 's
web UI.

* Visit this link:
  https://fedora.zanata.org/iteration/view/python-fedora

Compiling Message Catalogs
==========================

Message catalogs can be compiled for testing and should always be compiled
just prior to release.  Do this with the following script::
  python releaseutils.py build_catalogs

Compiled message catalogs should not be committed to source control.

Installing Message Catalogs
===========================

``python releaseutils.py install_catalogs`` will install the catalogs.  This
command may be customized through the use of environment variables.  See ``python
releaseutils.py --help`` for details.

-------
Release
-------

0) Commit all features, hotfixes, etc that you want in the release into the
   develop branch.

1) Checkout a copy of the repository and setup git flow::
     git clone git@github.com:fedora-infra/python-fedora.git
     cd python-fedora
     git flow init

2) Create a release branch for all of our work

     git flow release start $VERSION

3) Download new translations and verify they are valid by compiling them::
     zanata-cli pull
     python releaseutils.py build_catalogs
     # If everything checks out
     git commit -m 'Merge new translations from fedora.zanata.org'

4) Make sure that the NEWS file is accurate (use git log if needed).

5) Update python-fedora.spec and fedora/release.py with the new version
   information.
     # Make edits to python-fedora.spec and release.py
     git commit

6) Make sure the docs are proper and publish them::
     # Build docs and check for errors
     python setup.py build_sphinx
     # pypi
     python setup.py upload_docs

7) Push the release branch to the server::
     # Update files
     git flow release publish $VERSION

8) Go to a temporary directory and checkout a copy of the release::
     cd ..
     git clone git@github.com:fedora-infra/python-fedora.git release
     cd release
     git checkout release/$VERSION

9) Create the tarball in this clean checkout::
     python setup.py sdist

10) copy the dist/python-fedora-VERSION.tar.gz and python-fedora.spec files to
    where you build Fedora RPMS.  Do a test build::
     cp dist/python-fedora-*.tar.gz python-fedora.spec /srv/git/python-fedora/
     pushd /srv/git/python-fedora/
     fedpkg switch-branch master
     make mockbuild

11) Make sure the build completes.  Run rpmlint on the results.  Install and
    test the new packages::
     rpmlint *rpm
     sudo rpm -Uvh *noarch.rpm
     [test]

12) When satisfied that the build works, create a fresh tarball and upload to
    pypi::
     popd   # Back to the release directory
     python setup.py sdist upload --sign

13) copy the same tarball to fedorahosted.  The directory to upload to is
    slightly different for fedorahosted admins vs normal fedorahosted users:
    Admin::
      scp dist/python-fedora*tar.gz* fedorahosted.org:/srv/web/releases/p/y/python-fedora/
    Normal contributor::
      scp dist/python-fedora*tar.gz* fedorahosted.org:python-fedora

14) mark the release as finished in git::
     cd ../python-fedora
     git flow release finish $VERSION
     git push --all
     git push --tags

15) Finish building and pushing packages for Fedora.


NEWS
====
NEWS
====

:Authors: Toshio Kuratomi, Luke Macken, Ricky Elrod, Patrick Uiterwijk, Ralph Bean
:Version: 1.1.1

------
1.1.1
------

Bugfixes:

* Membership of the "signed_fpca" group was not always requested.
* Handle the requested group list being passed as a set (wink wink Pagure).


------
1.1.0
------

New features:

* Support for the new AAA system, which adds users who have signed agreement
  "foobar" to the ``signed_foobar`` group. Make use of this to set the
  ``cla_done`` key.


------
1.0.0
------
Backwards incompatible:

* Removed fedora.client.bodhi.
* Drop support for Django
* Drop support for TurboGears1
* Drop support for TurboGears2
* Drop support for flask_fas

New features:

* OpenIDCBaseClient now supports a client_secret argument
* Add support for PUT and DELETE requests to OpenIdBaseClient 'send_request' method.
* Make flask_fas_openid compatible with flask >= 1.0

Bugfixes

* OpenIdBaseClient The encoding arg of json.loads() is ignored and deprecated


------
0.10.0
------

* Seperate init_app in flask_fas_openid
* Some py3 compatibility
* Harden url() function

-----
0.9.0
-----

* Deprecated fedora.client.bodhi in favor of bodhi.client.bindings.
* Added OpenID Connect client
* Added file permission checks on openidbaseclient cookies


-----
0.8.0
-----

* Removed python 2.5 compatibility code
* Various python 3 fixes
* Documentation fixes
* OpenID endpoint check
* Cookie clearing on login in openidbaseclient
* Removed flask_fas backwards compatibility module

-----
0.7.1
-----

* Fix a regression in BodhiClient config parsing introduced in 0.7.0.

-----
0.7.0
-----

* Removed the pkgdb1 client.  The server has been gone for over a year now.
* Removed the flask_fas server component in favor of flask_fas_openid.
* Updated flask_fas_openid to forward along ssh and gpg public key information.
* Python3 compat enhancements from Craig Rodrigues and Ville Skyttä.

-----
0.6.4
-----

* Bodhi2Client query: document new _before query params
* Also perform request retries for failed POST requests.

-----
0.6.3
-----

* Adjust download URL for pypi.

-----
0.6.2
-----

* Force the Bodhi2 url in the BodhiClient shim (for /usr/bin/bodhi).

-----
0.6.1
-----

* Fix a bug in the BodhiClient where the Bodhi2Client would try to use the old
  Bodhi1 base url.
* Raise an AuthError if no username or password is provided to the openid login
  function.

-----
0.6.0
-----

* Cache session cookies between use for clients using OpenID.
* Retry failed requests for clients using OpenID
* Remove Bodhi server version detection code.
* Python3 support for the BodhiClient.
* Drop support for python-requests-1.x

-----
0.5.6
-----

* Python2.6 fix for string formatting in the BodhiClient.

-----
0.5.5
-----

* Consistent exception handling in the Bodhi2Client.
* More adjustments to bodhi server version detection.

-----
0.5.4
-----

* Adjust usage of python-six to be compatible with older versions on rhel7.
* More updates to detection of the bodhi server version.

-----
0.5.3
-----

* Better bodhi compatibility with 'fedpkg update'.

-----
0.5.2
-----

* Better version checking against the bodhi server.

-----
0.5.1
-----

* Typofix in the BodhiClient

-----
0.5.0
-----

* Bodhi2 compatibility!

* Ensure that the OpenIdBaseClient returns a munch object like the old
  BaseClient did.

* Minor fixes to some python3 compatibility.

-----
0.4.0
-----

* Port most modules to a python2/python3 compatible codebase!  Hooray!

* Deprecate the wiki client (it will be removed in the next major release).

* Switch over to using zanata for translations.

------
0.3.37
------

* Make the fedora_session file unhidden

* Instead of returning the JSON returned by fedoauth/ipsilon we return
  the response from the application when we sent to it the data returned
  by fedoauth/ipsilon.

* Update OpenIdProxyClient to allow specifying headers when sending a request

* Update bugzilla exception list.


------
0.3.36
------

* Fix the OpenId BaseClient NullHandler for logging

------
0.3.35
------

* Add Openid BaseClient for talking to our new web services that are using
  openid to auth.
* flask_fas_openid can return the json data from the openid request instead of
  the html page if requested.

* Add a requirement on urllib3 in setup.py so that pip will install it.

------
0.3.34
------

* **SECURITY** Fix the jsglobals.html template to escape usernames and human
  names to fix potential XSS flaw.
* **SECURITY** Make flask-fas-openid check that the page it is returning the
  user to after authentication is from within the application.

* Updated libravatar lookups to use the user's openid identifier instead of
  their email address.

------
0.3.33
------

* Minimum version of python is now 2.6.  Some things may work on older python
  but this is not guaranteed.  Patches to make it work may or may not be
  accepted.
* Added flask_fas_openid for the migration to OpenID
  This adds the requirements for python-openid, python-openid-teams and
  python-openid-cla.
* Rename the FAS_HTTPS_REQUIRED config variable in the flask_fas auth adapter
  to FAS_FLASK_COOKIE_REQUIRES_HTTPS.  The longer name is more cumbersome but
  the previous name was too often mistaken to be setting whether we needed to
  connect to the fas server with https when it really was about whether the
  session cookie should require https to be returned to the server (the
  cookie's secure flag).
* Change the groups attribute of the flask_fas and flask_fas_openid auth
  adapters to be a frozenset instead of a list
* Set a configurable timeout on how long a client request will wait for the
  server to setup a connection in BaseClient and ProxyClient.
* Fix instance-wide default retries in BaseClient (and subclasses) being
  overridden by the default value of retries in BaseClient.send_request()
* Fix a performance bug in the client modules when the server returns large
  json strings.  python-requests defaults to trying to detect the charset of

pavement.py
try:
    # 1.0
    from paver.easy import path as paver_path
    from paver.easy import sh as paver_sh
    from paver.easy import *
    import paver.misctasks
    from paver import setuputils
    setuputils.install_distutils_tasks()
    PAVER_VER = '1.0'
except ImportError:
    # 0.8
    from paver.defaults import *
    from paver.runtime import path as paver_path
    from paver.runtime import sh as paver_sh
    PAVER_VER = '0.8'

import sys, os
import glob
import paver.doctools
from setuptools import find_packages, command

sys.path.insert(0, str(paver_path.getcwd()))

from fedora.release import *

options(
    setup = Bunch(
        name=NAME,
        version=VERSION,
        description=DESCRIPTION,
        author=AUTHOR,
        author_email=EMAIL,
        license=LICENSE,
        keywords='Fedora Python Webservices',
        url=URL,
        download_url=DOWNLOAD_URL,
        packages=find_packages(),
        py_modules=['flask_fas', 'flask_fas_openid'],
        include_package_data=True,
        # non-setuptools package.  When everything we care about uses
        # python-2.5 distutils we can add these:
        #   for bodhi (koji yum)
        #   pycurl
        install_requires=['bunch', 'kitchen', 'requests'],
        # And these to extra_require:
        #   for widgets: (bugzilla feedparser)
        extras_require = {
            'tg' : ['TurboGears >= 1.0.4', 'SQLAlchemy', 'decorator'],
            'wsgi': ['repoze.who', 'Beaker', 'Paste'],
            'flask': ['Flask', 'Flask_WTF', 'openid'],
            },
        #test_require = ['TurboGears >= 2.0', 'nose', ],
        entry_points = {
            'turbogears.identity.provider' : (
                'jsonfas = fedora.tg.identity.jsonfasprovider1:JsonFasIdentityProvider [tg]',
                'jsonfas2 = fedora.tg.identity.jsonfasprovider2:JsonFasIdentityProvider [tg]',
                'sqlobjectcsrf = fedora.tg.identity.soprovidercsrf:SqlObjectCsrfIdentityProvider [tg]'),
            'turbogears.visit.manager' : (
                'jsonfas = fedora.tg.visit.jsonfasvisit1:JsonFasVisitManager [tg]',
                'jsonfas2 = fedora.tg.visit.jsonfasvisit2:JsonFasVisitManager [tg]'),
            # Needed for the test suite
            'paste.app_factory' : (
                'main = fedora.wsgi.test.testapp:make_app'),
            'paste.app_install' : (
                'main = pylons.util:PylonsInstaller'),
            'paste.filter_app_factory': (
                'faswho = fedora.wsgi.faswho:make_faswho_middleware')
            },
        message_extractors = {
            'fedora': [('**.py', 'python', None),
                ('tg2/templates/mako/**.mak', 'mako', None),
                ('tg2/templates/genshi/**.html', 'mako', None),
                ('tg/templates/genshi/**.html', 'genshi', None),],
            },
        classifiers = [
            'Development Status :: 4 - Beta',
            'Framework :: TurboGears',
            'Framework :: Django',
            'Intended Audience :: Developers',
            'Intended Audience :: System Administrators',
            'License :: OSI Approved :: GNU Lesser General Public License v2 or later (LGPLv2+)',
            'Programming Language :: Python :: 2.5',
            'Programming Language :: Python :: 2.6',
            'Programming Language :: Python :: 2.7',
            'Topic :: Internet :: WWW/HTTP',
            'Topic :: Software Development :: Libraries :: Python Modules',
            ],
        ),
    minilib=Bunch(
        extra_files=['doctools']
        ),
    sphinx=Bunch(
        docroot='.',
        builddir='build-doc',
        sourcedir='doc'
        ),
    pylint=Bunch(
        module=['fedora']
        ),
    publish=Bunch(
        doc_location='fedorahosted.org:/srv/web/releases/p/y/python-fedora/doc/',
        tarball_location='fedorahosted.org:/srv/web/releases/p/y/python-fedora/'
        ),
    i18n=Bunch(
        builddir='locale',
        installdir='/usr/share/locale',
        domain='python-fedora',
        ),
    ### FIXME: These are due to a bug in paver-1.0
    # http://code.google.com/p/paver/issues/detail?id=24
    sdist=Bunch(),
    )

@task
@needs(['html'])
def publish_doc():
    options.order('publish', add_rest=True)
    command = 'rsync -av build-doc/html/ %s' % (options.doc_location,)
    if PAVER_VER == '0.8':
        command = [command]
    dry(command, paver_sh, command)

@task
@needs(['sdist'])
def publish_tarball():
    options.order('publish', add_rest=True)
    tarname = '%s-%s.tar.gz' % (options.name, options.version)
    command = 'scp dist/%s %s' % (tarname, options.tarball_location)
    if PAVER_VER == '0.8':
        command = [command]
    dry(command, paver_sh, command)

@task
@needs(['publish_doc', 'publish_tarball'])
def publish():
    pass

try:
    import babel.messages.frontend
    has_babel = True
except ImportError:
    has_babel = False

if has_babel and PAVER_VER != '0.8':
    @task
    def make_catalogs():
        '''Compile all message catalogs for release'''
        options.order('i18n', add_rest=True)
        for po_file in glob.glob('translations/*.po'):
            locale, ext = os.path.splitext(os.path.basename(po_file))
            build_dir = paver_path(os.path.join(options.builddir, locale,
                'LC_MESSAGES'))

            try:
                build_dir.makedirs(mode=0755)
            except OSError, e:
                # paver < 1.0 raises if directory exists
                if e.errno == 17:
                    pass
                else:
                    raise
            if 'compile_catalog' in options.keys():
                defaults = options.compile_catalog
            else:
                defaults = Bunch(domain=options.domain,
                        directory=options.builddir)
                options.compile_catalog = defaults

            defaults.update({'input-file': po_file, 'locale': locale})
            ### FIXME: compile_catalog cannot handle --dry-run on its own
            dry('paver compile_catalog -D %(domain)s -d %(directory)s'
                    ' -i %(input-file)s --locale %(locale)s' % defaults,
                    paver_sh, 'paver compile_catalog -D %(domain)s' \
                        ' -d %(directory)s -i %(input-file)s' \
                        ' --locale %(locale)s' % defaults)
            ### FIXME: Need to get call_task to call this repeatedly
            # because options.compile_catalog has changed
            #dry('paver compile_catalog -D %(domain)s -d %(directory)s'
            #        ' -i %(input-file)s --locale %(locale)s' % defaults,
            #        call_task, 'babel.messages.frontend.compile_catalog', options)

#
# Backends
#

def _apply_root(args, path):
    '''Add the root value to the start of the path'''
    if 'root' in args:
        if path.startswith('/'):
            path = path[1:]
        path = paver_path(os.path.join(args['root'], path))
    else:
        path = paver_path(path)
    return path

def _install_catalogs(args):
    '''Install message catalogs in their proper location on the filesystem.

    Note: To use this with non-default commandline arguments, you must use 
    '''
    # Rebuild message catalogs
    if 'skip_build' not in args and 'skip-build' not in args:
        call_task('make_catalogs')

    options.order('i18n', add_rest=True)
    # Setup the install_dir
    if 'install_catalogs' in args:
        cat_dir = args['install_catalogs']
    elif 'install_data' in args:
        cat_dir = os.path.join(args['install_data'], 'locale')
    else:
        cat_dir = options.installdir

    # Setup the install_dir
    cat_dir = _apply_root(args, cat_dir)

    for catalog in paver_path(options.builddir).walkfiles('*.mo'):
        locale_dir = catalog.dirname()
        path = paver_path('.')
        for index, nextpath in enumerate(locale_dir.splitall()):
            path = path.joinpath(nextpath)
            if paver_path(options.builddir).samefile(path):
                install_locale = cat_dir.joinpath(os.path.join(
                        *locale_dir.splitall()[index + 1:]))
                try:
                    install_locale.makedirs(mode=0755)
                except OSError, e:
                    # paver < 1.0 raises if directory exists
                    if e.errno == 17:
                        pass
                    else:
                        raise
                install_locale = install_locale.joinpath(catalog.basename())
                if install_locale.exists():
                    install_locale.remove()
                dry('cp %s %s'%  (catalog, install_locale),
                        catalog.copy, install_locale)
                dry('chmod 0644 %s'%  install_locale,
                        install_locale.chmod, 0644)

@task
@cmdopts([('root=', None, 'Base root directory to install into'),
    ('install-catalogs=', None, 'directory that locale catalogs go in'),
    ('skip-build', None, 'Skip directly to installing'),
    ])
def install_catalogs():
    _install_catalogs(options.install_catalogs)
    pass

@task
@needs(['make_catalogs', 'setuptools.command.sdist'])
def sdist():
    pass

### FIXME: setuptools.command.install does not respond to --dry-run
if PAVER_VER != '0.8':
    # Paver 0.8 will have to explicitly install the message catalogs
    # As overriding setuptools.command gives errors on paver-0.8
    @task
    @needs(['setuptools.command.install'])
    def install():
        '''Override the setuptools install.'''
        _install_catalogs(options.install)

#
# Generic Tasks
#

try:
    from pylint import lint
    has_pylint = True
except ImportError:
    has_pylint = False

if has_pylint:
    @task
    def pylint():
        '''Check the module you're building with pylint.'''
        options.order('pylint', add_rest=True)
        pylintopts = options.module
        dry('pylint %s' % (" ".join(pylintopts)), lint.Run, pylintopts)


@task
def test():
    sh("nosetests --where=fedora/wsgi/test")
